### ISSUE TITLE
[core][jadx-next] Wrong class name used for package-private classes if source filename is different

### ISSUE THREAD

**Issue by nitram84** [skylot/jadx#1795]
Steps to reproduce:
- example compiled with openjdk8 and decompiled with use-dx option
- tested with latest git version (jadx-next)
- reproducible also on master, but here deobfuscation has to be activated, even in cases when deobfucation is not needed (testet with longer names).

Filename "a.java"

```java
class b {
}
```

Same with smali:

```smali
.class Lb;
.super Ljava/lang/Object;
.source "a.java"

# direct methods
.method constructor <init>()V
    .registers 1

    .prologue
    .line 1
    invoke-direct {p0}, Ljava/lang/Object;-><init>()V

    return-void
.end method
```

The decompiled result is:

```java
/* renamed from: b */
/* loaded from: [...] */
class a {
    a() {
    }
}
```

If deobfuscation is activated, affected classes aren't deobfuscated. At least the filenames are correct, when sources are exported.

But there is also a special case: The name of a source file in my example could be "package-info.java". In this case the class name will be correct, but a wrong filename is used for exporting the source.

There are multiple issues with "package-info". I should open a new ticket for this.

**Comment by nitram84** [comment]
Hi @skylot. Thank you for your fix in jadx-next.

I'm not sure whether your fix is correct. In my opinion this issue has nothing to do with deobfuscation (except of master branch).

For private or package-private classes the source name can differ from the real class name. Class names might be deobfuscated, but I think the filename should never be used as the new class name in any case. In my option for each class the original source name should be stored and should only be used to export sources. Classes should be aggregated to source files at least for private or package-private classes. Perhaps primary entity for export should be source file and not class node, because source file can contain more than one class. I would ignore kotlin clases and others for now.

Here is an example how I would exploit this issue:

```java
// Fake source file: JadxDecompilerDetected.java
package detectjadx;

class ThisIsTheRealClassname {
  public Throwable getThrowable() {
    return new Throwable();
  }
}
```

```java
// Source file: RuntimeDecompilerDetector.java
package detectjadx;

public class RuntimeDecompilerDetector {

  public static class DecompilerDetectedExcetion extends Exception {
  }

  public static void detectDecompiler() throws DecompilerDetectedExcetion {
    try {
      // check if jadx renamed the test class
      RuntimeDecompilerDetector.class.getClassLoader().loadClass("detectjadx.JadxDecompilerDetected");
      throw new DecompilerDetectedExcetion();
    } catch (ClassNotFoundException e) {
      // check if jadx used the expected filename for export
      try {
        Class<ThisIsTheRealClassname> clazz = (Class<ThisIsTheRealClassname>) RuntimeDecompilerDetector.class
            .getClassLoader().loadClass("detectjadx.ThisIsTheRealClassname");
        ThisIsTheRealClassname test = clazz.newInstance();
        if (!"JadxDecompilerDetected.java".equals(test.getThrowable().getStackTrace()[0].getFileName())) {
          throw new DecompilerDetectedExcetion();
        }
        
        // TODO Do repeat last check with a kotlin class if jadx could generate kotlin files
      } catch (ClassNotFoundException | InstantiationException | IllegalAccessException e1) {
        throw new DecompilerDetectedExcetion();
      }

      // decompiler check passed, everything is ok
    }
  }

  public static void main(String[] args) {
    try {
      RuntimeDecompilerDetector.detectDecompiler();
    } catch (DecompilerDetectedExcetion e) {
      // Do some bad things
      System.out.println(
          "You are using a decompiled version of this app. Your account has been locked. This incident will be reported!");
    }
  }
}
```

As long as the source name of my canary class is different/fake, it would be possible to detect a decompiled and recompiled app, no matter of obfuscation or deobfuscation. I think this can be prevented with a fix before someone else has the same idea.

**Comment by skylot** [comment]
@nitram84 I understand now what you're trying to do.
But from my experience source name in most cases is literally garbage and doesn't contain any useful info, thus this feature (`Use source file name as class name alias`) is disabled by default.
Also in most cases decompiled source will not compile back and you should use apktool and edit smali code, so your use case is quite rare. And for such rare cases you can use jadx script to fit your needs.

**Comment by nitram84** [comment]
@skylot Thank you for your answer. But I'd like to stay with jadx. I did some experiments with post processing of jadx outputs: dependency detection, auto-resource-cleaning, legacy-http-client detection, ... and I can say, first recompilable apps out-of-the-box are possible, but still a lot of work is needed. I have a small list of apps I recompiled successfully with jadx. But step by step, there are still a lot of gradle and resource issues.

In my opinion the most significant issue that prevents decompiling is https://github.com/skylot/jadx/issues/1073 (reproducible with almost any app, to fix this \src\main\res\values\attrs.xml has to be decoded first and its flag and enum definitions should be added to the attrib map in ManifestAttributes).

I'll soon prepare some PRs for some other issues I found during my tests.
